File: server/internal/httpapi/idempotency.go
package httpapi

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const (
	idempotencyTTL     = 24 * time.Hour
	idempotencyLease   = 2 * time.Minute
	maxReplayBodyBytes = 1024 * 1024
)

// IdempotencyDecision tells a create handler whether to execute or replay.
type IdempotencyDecision int

const (
	IdempotencyProceed IdempotencyDecision = iota
	IdempotencyReplay
	IdempotencyConflict
	IdempotencyInProgress
)

// ActorScope is the complete idempotency namespace.
type ActorScope struct {
	ActorType     string
	ActorID       uuid.UUID
	Method        string
	CanonicalPath string
}

// StoredResponse is the safe replay payload retained for 24 hours.
type StoredResponse struct {
	Status  int
	Headers map[string][]string
	Body    []byte
}

// IdempotencyResult is returned by Store.Begin.
type IdempotencyResult struct {
	Decision IdempotencyDecision
	ClaimID  uuid.UUID
	Response StoredResponse
}

// IdempotencyStore supports actor-scoped claim, replay, and completion.
type IdempotencyStore interface {
	Begin(
		context.Context,
		ActorScope,
		string,
		[32]byte,
		time.Time,
	) (IdempotencyResult, error)
	Complete(context.Context, uuid.UUID, StoredResponse, time.Time) error
	Abandon(context.Context, uuid.UUID) error
}

// PostgresIdempotencyStore persists claims in the authoritative database.
type PostgresIdempotencyStore struct {
	Pool *pgxpool.Pool
}

// ValidateIdempotencyKey enforces the public 16-128 character header contract.
func ValidateIdempotencyKey(key string) error {
	if !utf8.ValidString(key) || len(key) < 16 || len(key) > 128 {
		return errors.New("Idempotency-Key must contain 16 to 128 characters")
	}
	for _, character := range key {
		if character < 0x21 || character > 0x7e {
			return errors.New("Idempotency-Key must contain visible ASCII characters")
		}
	}
	return nil
}

// FingerprintJSON hashes a canonical JSON representation so harmless object
// whitespace and key ordering do not create false conflicts.
func FingerprintJSON(body []byte) ([32]byte, error) {
	var value any
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.UseNumber()
	if err := decoder.Decode(&value); err != nil {
		return [32]byte{}, errors.New("request body is not valid JSON")
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return [32]byte{}, errors.New("request body contains multiple JSON values")
	}
	canonical, err := json.Marshal(value)
	if err != nil {
		return [32]byte{}, errors.New("canonicalize request body")
	}
	return sha256.Sum256(canonical), nil
}

// Begin atomically claims a key or returns its prior result/conflict.
func (store PostgresIdempotencyStore) Begin(
	ctx context.Context,
	scope ActorScope,
	key string,
	fingerprint [32]byte,
	now time.Time,
) (IdempotencyResult, error) {
	if store.Pool == nil {
		return IdempotencyResult{}, errors.New("idempotency store is unavailable")
	}
	if err := validateScope(scope); err != nil {
		return IdempotencyResult{}, err
	}
	if err := ValidateIdempotencyKey(key); err != nil {
		return IdempotencyResult{}, err
	}
	pathHash := sha256.Sum256([]byte(scope.CanonicalPath))
	tx, err := store.Pool.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return IdempotencyResult{}, errors.New("begin idempotency claim")
	}
	defer func() {
		_ = tx.Rollback(context.Background())
	}()

	if _, err := tx.Exec(
		ctx,
		`DELETE FROM idempotency_records
		 WHERE actor_type = $1 AND actor_id = $2 AND method = $3
		   AND canonical_path_hash = $4 AND idempotency_key = $5
		   AND expires_at <= $6`,
		scope.ActorType,
		scope.ActorID,
		scope.Method,
		pathHash[:],
		key,
		now,
	); err != nil {
		return IdempotencyResult{}, errors.New("expire idempotency claim")
	}

	claimID := uuid.New()
	err = tx.QueryRow(
		ctx,
		`INSERT INTO idempotency_records (
			id, actor_type, actor_id, method, canonical_path,
			canonical_path_hash, idempotency_key, fingerprint,
			lease_expires_at, expires_at
		 ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
		 ON CONFLICT DO NOTHING
		 RETURNING id`,
		claimID,
		scope.ActorType,
		scope.ActorID,
		scope.Method,
		scope.CanonicalPath,
		pathHash[:],
		key,
		fingerprint[:],
		now.Add(idempotencyLease),
		now.Add(idempotencyTTL),
	).Scan(&claimID)
	switch {
	case err == nil:
		if err := tx.Commit(ctx); err != nil {
			return IdempotencyResult{}, errors.New("commit idempotency claim")
		}
		return IdempotencyResult{
			Decision: IdempotencyProceed,
			ClaimID:  claimID,
		}, nil
	case !errors.Is(err, pgx.ErrNoRows):
		return IdempotencyResult{}, errors.New("create idempotency claim")
	}

	var (
		existingID     uuid.UUID
		existingPrint  []byte
		status         *int
		headersJSON    []byte
		responseBody   []byte
		leaseExpiresAt time.Time
	)
	if err := tx.QueryRow(
		ctx,
		`SELECT id, fingerprint, response_status,
		        COALESCE(response_headers, '{}'::jsonb), response_body, lease_expires_at
		   FROM idempotency_records
		  WHERE actor_type = $1 AND actor_id = $2 AND method = $3
		    AND canonical_path_hash = $4 AND idempotency_key = $5
		  FOR UPDATE`,
		scope.ActorType,
		scope.ActorID,
		scope.Method,
		pathHash[:],
		key,
	).Scan(
		&existingID,
		&existingPrint,
		&status,
		&headersJSON,
		&responseBody,
		&leaseExpiresAt,
	); err != nil {
		return IdempotencyResult{}, errors.New("read idempotency claim")
	}

	if !bytes.Equal(existingPrint, fingerprint[:]) {
		if err := tx.Commit(ctx); err != nil {
			return IdempotencyResult{}, errors.New("commit idempotency conflict")
		}
		return IdempotencyResult{Decision: IdempotencyConflict}, nil
	}
	if status != nil {
		headers := make(map[string][]string)
		if err := json.Unmarshal(headersJSON, &headers); err != nil {
			return IdempotencyResult{}, errors.New("decode idempotency response")
		}
		if err := tx.Commit(ctx); err != nil {
			return IdempotencyResult{}, errors.New("commit idempotency replay")
		}
		return IdempotencyResult{
			Decision: IdempotencyReplay,
			Response: StoredResponse{
				Status:  *status,
				Headers: headers,
				Body:    responseBody,
			},
		}, nil
	}
	if !leaseExpiresAt.After(now) {
		if _, err := tx.Exec(
			ctx,
			`UPDATE idempotency_records
			    SET lease_expires_at = $2
			  WHERE id = $1`,
			existingID,
			now.Add(idempotencyLease),
		); err != nil {
			return IdempotencyResult{}, errors.New("renew idempotency claim")
		}
		if err := tx.Commit(ctx); err != nil {
			return IdempotencyResult{}, errors.New("commit idempotency claim renewal")
		}
		return IdempotencyResult{
			Decision: IdempotencyProceed,
			ClaimID:  existingID,
		}, nil
	}
	if err := tx.Commit(ctx); err != nil {
		return IdempotencyResult{}, errors.New("commit idempotency in-progress result")
	}
	return IdempotencyResult{Decision: IdempotencyInProgress}, nil
}

// Complete stores replayable responses and drops 5xx claims.
func (store PostgresIdempotencyStore) Complete(
	ctx context.Context,
	claimID uuid.UUID,
	response StoredResponse,
	now time.Time,
) error {
	if store.Pool == nil {
		return errors.New("idempotency store is unavailable")
	}
	if response.Status >= 500 {
		return store.Abandon(ctx, claimID)
	}
	if response.Status < 100 || response.Status > 499 {
		return errors.New("idempotency response status is invalid")
	}
	if len(response.Body) > maxReplayBodyBytes {
		return fmt.Errorf("idempotency response exceeds %d bytes", maxReplayBodyBytes)
	}
	headers := replayHeaders(response.Headers)
	encoded, err := json.Marshal(headers)
	if err != nil {
		return errors.New("encode idempotency response headers")
	}
	tag, err := store.Pool.Exec(
		ctx,
		`UPDATE idempotency_records
		    SET response_status = $2,
		        response_headers = $3::jsonb,
		        response_body = $4,
		        completed_at = $5,
		        lease_expires_at = $5,
		        expires_at = $6
		  WHERE id = $1 AND response_status IS NULL`,
		claimID,
		response.Status,
		string(encoded),
		response.Body,
		now,
		now.Add(idempotencyTTL),
	)
	if err != nil {
		return errors.New("complete idempotency claim")
	}
	if tag.RowsAffected() != 1 {
		return errors.New("idempotency claim is no longer active")
	}
	return nil
}

// Abandon removes an incomplete claim so a failed operation can be retried.
func (store PostgresIdempotencyStore) Abandon(
	ctx context.Context,
	claimID uuid.UUID,
) error {
	if store.Pool == nil {
		return errors.New("idempotency store is unavailable")
	}
	if _, err := store.Pool.Exec(
		ctx,
		"DELETE FROM idempotency_records WHERE id = $1 AND response_status IS NULL",
		claimID,
	); err != nil {
		return errors.New("abandon idempotency claim")
	}
	return nil
}

// Replay writes a previously stored response.
func Replay(response http.ResponseWriter, stored StoredResponse) {
	for name, values := range replayHeaders(stored.Headers) {
		for _, value := range values {
			response.Header().Add(name, value)
		}
	}
	response.Header().Set("Idempotency-Replayed", "true")
	response.WriteHeader(stored.Status)
	_, _ = response.Write(stored.Body)
}

func validateScope(scope ActorScope) error {
	if scope.ActorType != "user" && scope.ActorType != "agent" {
		return errors.New("idempotency actor type is invalid")
	}
	if scope.ActorID == uuid.Nil || scope.Method != strings.ToUpper(scope.Method) ||
		scope.Method == "" || len(scope.CanonicalPath) == 0 ||
		len(scope.CanonicalPath) > 2048 ||
		!strings.HasPrefix(scope.CanonicalPath, "/") {
		return errors.New("idempotency scope is invalid")
	}
	return nil
}

func replayHeaders(headers map[string][]string) map[string][]string {
	allowed := map[string]struct{}{
		"Content-Type": {},
		"Location":     {},
		"Etag":         {},
	}
	result := make(map[string][]string)
	for name, values := range headers {
		canonical := http.CanonicalHeaderKey(name)
		if _, ok := allowed[canonical]; !ok {
			continue
		}
		result[canonical] = append([]string(nil), values...)
	}
	return result
}

File: server-ts/src/http/idempotency.ts
import { createHash, randomUUID } from 'node:crypto';
import type { Sql } from '../db/pool.ts';
import { canonicalJSON, parseWithRawNumbers } from './canonical-json.ts';

/**
 * Durable idempotency claims.
 *
 * A claim is taken before the work runs and completed with the response, so a
 * retried request replays what the first one produced rather than doing it
 * twice. That matters most where the work is not undoable — dispatching a run,
 * charging for tokens, posting to a third party.
 *
 * The claim is a row, not a lock: it survives a restart, which an in-memory
 * map would not, and the window it protects can outlive the process.
 */

const LEASE_MS = 2 * 60 * 1000;
const TTL_MS = 24 * 60 * 60 * 1000;
const MAX_REPLAY_BODY = 1024 * 1024;

export type Decision = 'proceed' | 'replay' | 'conflict' | 'in-progress';

/** The complete namespace a key is unique within. */
export interface ActorScope {
   actorType: 'user' | 'agent';
   actorId: string;
   method: string;
   canonicalPath: string;
}

export interface StoredResponse {
   status: number;
   headers: Record<string, string[]>;
   body: string;
}

export interface ClaimResult {
   decision: Decision;
   claimId?: string;
   response?: StoredResponse;
}

export function validateIdempotencyKey(key: string): boolean {
   // Bytes, matching Go's len() on a string: a 128-character key of multi-byte
   // characters is longer than 128 bytes and is refused there too.
   const bytes = Buffer.byteLength(key, 'utf8');
   return bytes >= 16 && bytes <= 128 && /^[\x21-\x7e]+$/.test(key);
}

/**
 * Hashes a canonical form of the body, so whitespace and key order do not
 * make the same request look like a different one.
 *
 * The canonical form is not JavaScript's — see `canonical-json.ts`. It cannot
 * be: fingerprints of past requests are already stored, and a form that hashed
 * them differently would turn a legitimate replay into an IDEMPOTENCY_CONFLICT.
 *
 * Throws when the body is not exactly one JSON value — a caller sending
 * something unparseable has not made a request worth claiming a key for.
 */
export function fingerprintJSON(body: string): Buffer {
   return createHash('sha256').update(canonicalJSON(parseWithRawNumbers(body))).digest();
}

function validScope(scope: ActorScope): boolean {
   return (
      (scope.actorType === 'user' || scope.actorType === 'agent') &&
      scope.actorId !== '' &&
      scope.method !== '' &&
      scope.method === scope.method.toUpperCase() &&
      scope.canonicalPath.startsWith('/') &&
      scope.canonicalPath.length <= 2048
   );
}

/**
 * Only headers that are safe and meaningful to repeat.
 *
 * A replay must not resurrect the original's Set-Cookie or request id — the
 * body is the same answer, but the transport is a new exchange.
 */
const REPLAYABLE = new Set(['content-type', 'location', 'etag']);

export function replayHeaders(headers: Record<string, string[]>): Record<string, string[]> {
   const result: Record<string, string[]> = {};
   for (const [name, values] of Object.entries(headers)) {
      if (REPLAYABLE.has(name.toLowerCase())) result[canonical(name)] = [...values];
   }
   return result;
}

/** Headers are stored canonicalised, so a replay fingerprints to the same bytes. */
function canonical(name: string): string {
   return name
      .split('-')
      .map((part) => (part ? part[0]!.toUpperCase() + part.slice(1).toLowerCase() : part))
      .join('-');
}

export class IdempotencyStore {
   private readonly sql: Sql;
   private readonly clock: () => Date;
   private readonly newId: () => string;

   constructor(sql: Sql, clock: () => Date = () => new Date(), newId = randomUUID) {
      this.sql = sql;
      this.clock = clock;
      this.newId = newId;
   }

   /**
    * Claims a key, or reports what the key already means.
    *
    * Four outcomes, and the distinction between the last two is the point:
    * a completed claim replays its stored response, while one still inside its
    * lease is another request in flight — answered 409 with Retry-After rather
    * than executed a second time.
    */
   async begin(scope: ActorScope, key: string, fingerprint: Buffer): Promise<ClaimResult> {
      if (!validScope(scope)) throw new Error('idempotency scope is invalid');
      if (!validateIdempotencyKey(key)) throw new Error('idempotency key is invalid');

      const now = this.clock();
      const nowIso = now.toISOString();
      const pathHash = createHash('sha256').update(scope.canonicalPath).digest();

      return this.sql.begin(async (tx) => {
         // An expired claim is removed first, so the key becomes usable again
         // rather than conflicting forever with a record nobody can replay.
         await tx`
            DELETE FROM idempotency_records
             WHERE actor_type = ${scope.actorType} AND actor_id = ${scope.actorId}
               AND method = ${scope.method} AND canonical_path_hash = ${pathHash}
               AND idempotency_key = ${key} AND expires_at <= ${nowIso}`;

         const claimId = this.newId();
         const inserted = await tx`
            INSERT INTO idempotency_records (
               id, actor_type, actor_id, method, canonical_path,
               canonical_path_hash, idempotency_key, fingerprint,
               lease_expires_at, expires_at
            ) VALUES (
               ${claimId}, ${scope.actorType}, ${scope.actorId}, ${scope.method},
               ${scope.canonicalPath}, ${pathHash}, ${key}, ${fingerprint},
               ${new Date(now.getTime() + LEASE_MS).toISOString()},
               ${new Date(now.getTime() + TTL_MS).toISOString()}
            )
            ON CONFLICT DO NOTHING
            RETURNING id`;
         if (inserted.length > 0) {
            return { decision: 'proceed' as const, claimId: inserted[0]!.id as string };
         }

         const [existing] = await tx`
            SELECT id, fingerprint, response_status,
                   COALESCE(response_headers, '{}'::jsonb) AS response_headers,
                   response_body, lease_expires_at
              FROM idempotency_records
             WHERE actor_type = ${scope.actorType} AND actor_id = ${scope.actorId}
               AND method = ${scope.method} AND canonical_path_hash = ${pathHash}
               AND idempotency_key = ${key}
             FOR UPDATE`;
         if (!existing) throw new Error('idempotency claim vanished');

         // The same key with a different body is a caller bug: answering with
         // the first response would silently discard the second request.
         if (!(existing.fingerprint as Buffer).equals(fingerprint)) {
            return { decision: 'conflict' as const };
         }

         if (existing.response_status !== null) {
            return {
               decision: 'replay' as const,
               response: {
                  status: existing.response_status as number,
                  headers: (existing.response_headers ?? {}) as Record<string, string[]>,
                  body: (existing.response_body as Buffer | null)?.toString('utf8') ?? '',
               },
            };
         }

         // A lease that has run out means the holder died mid-flight; take it
         // over rather than leaving the key unusable for its whole TTL.
         if (new Date(existing.lease_expires_at as string) <= now) {
            await tx`
               UPDATE idempotency_records
                  SET lease_expires_at = ${new Date(now.getTime() + LEASE_MS).toISOString()}
                WHERE id = ${existing.id as string}`;
            return { decision: 'proceed' as const, claimId: existing.id as string };
         }
         return { decision: 'in-progress' as const };
      });
   }

   /**
    * Stores the response for replay.
    *
    * A 5xx abandons the claim instead: the request failed for a reason that
    * may not recur, and pinning that failure for 24 hours would make a
    * transient fault permanent for anyone retrying with the same key.
    */
   async complete(claimId: string, response: StoredResponse): Promise<void> {
      if (response.status >= 500) return this.abandon(claimId);
      if (response.status < 100 || response.status > 499) {
         throw new Error('idempotency response status is invalid');
      }
      if (Buffer.byteLength(response.body, 'utf8') > MAX_REPLAY_BODY) {
         throw new Error(`idempotency response exceeds ${MAX_REPLAY_BODY} bytes`);
      }

      const now = this.clock();
      const updated = await this.sql`
         UPDATE idempotency_records
            SET response_status = ${response.status},
                response_headers = ${this.sql.json(replayHeaders(response.headers))}::jsonb,
                response_body = ${Buffer.from(response.body, 'utf8')},
                completed_at = ${now.toISOString()},
                lease_expires_at = ${now.toISOString()},
                expires_at = ${new Date(now.getTime() + TTL_MS).toISOString()}
          WHERE id = ${claimId} AND response_status IS NULL`;
      if (updated.count !== 1) throw new Error('idempotency claim is not open');
   }

   /** Releases an uncompleted claim so the key can be used again immediately. */
   async abandon(claimId: string): Promise<void> {
      await this.sql`
         DELETE FROM idempotency_records WHERE id = ${claimId} AND response_status IS NULL`;
   }
}
